source/extensions/common/matcher/matcher.cc

```diff
@@ -64,6 +105,53 @@ void buildMatcher(const envoy::config::common::matcher::v3::MatchPredicate& matc
   matchers[new_matcher->index()] = std::move(new_matcher);
 }
 
+// TODO(shivanshu21): Once HeaderMatcher is deprecated in favor of StringMatcher,
+// this overloaded method will no longer be required as we will use
+// envoy::config::common::matcher::v3::MatchPredicate
+void buildMatcher(const envoy::type::matcher::v3::MatchPredicate& match_config,
+                  std::vector<MatcherPtr>& matchers) {
+  matchers.emplace_back(nullptr);
+
+  MatcherPtr new_matcher;
+  switch (match_config.rule_case()) {
+  case envoy::type::matcher::v3::MatchPredicate::RuleCase::kOrMatch:
+    new_matcher = std::make_unique<SetLogicMatcher>(match_config.or_match(), matchers,
+                                                    SetLogicMatcher::Type::Or);
+    break;
+  case envoy::type::matcher::v3::MatchPredicate::RuleCase::kAndMatch:
+    new_matcher = std::make_unique<SetLogicMatcher>(match_config.and_match(), matchers,
+                                                    SetLogicMatcher::Type::And);
+    break;
+  case envoy::type::matcher::v3::MatchPredicate::RuleCase::kNotMatch:
+    new_matcher = std::make_unique<NotMatcher>(match_config.not_match(), matchers);
+    break;
+  case envoy::type::matcher::v3::MatchPredicate::RuleCase::kAnyMatch:
+    new_matcher = std::make_unique<AnyMatcher>(matchers);
+    break;
+  case envoy::type::matcher::v3::MatchPredicate::RuleCase::kHttpResponseHeadersMatch:
+    new_matcher = std::make_unique<HttpResponseHeadersMatcher>(
+        match_config.http_response_headers_match(), matchers);
+    break;
+  default:
+    NOT_REACHED_GCOVR_EXCL_LINE;
+  }
```

Comment:
@snowp is your new generic header definitely not going to be ready in time for this PR? The implementation and choice of having two almost identical matchers here is a lot of tech debt. If there's any chance we can benefit from your work it'd be much more desirable.

Reply:
Probably not with just me working on the generic header matching, between the other work that I'm doing and some scheduled OOO I think it will be some time before that lands, though I'm not sure what the deadline is for this PR.

If someone wants to help land the generic matching faster that would likely help, I would definitely prefer to avoid this additional tech debt if possible.